### ISSUE TITLE
Cancelation event not cached by server

### ISSUE THREAD

**Issue by watjurk** [improbable-eng/grpc-web#899]
### **Versions of relevant software used:**
@improbable-eng/grpc-web: v0.14.0 (go)
@improbable-eng/grpc-web: 0.14.0 (js)

### **What happened:**
Transport: Websocket.
When user makes `unary` or `server stream` request and cancels it, the cancelation event 
isn't cached by the server.

### **What you expected to happen:**
Server should cach the cancelation event.

### **How to reproduce it (as minimally and precisely as possible):**
Transport: Websocket.
Create `unary` or `server stream` service, cancel it on user side and try caching this cancel event on server.

### **My fix:**
webSocketWrappedReader in websocket_wrapper.go  should read one message ahead from websocket.
I've already implemented this in code, so if you would like to I could make a PR.

**Comment by johanbrandhorst** [comment]
What do you mean by caching the cancellation exactly? We have tests that test cancellation propagation.

**Comment by watjurk** [comment]
I meant receiving on context.Cancel(). Your test may not catch this because this error occurs only when transport is set to websocket.

**Comment by watjurk** [comment]
Hi, so what are your thoughts on this ? If you would like to I can provide example repo with this issue.

**Comment by johanbrandhorst** [comment]
The best solution would be in the form of a PR with a failing test case. You can find the cancellation tests here: https://github.com/improbable-eng/grpc-web/blob/master/integration_test/ts/src/cancellation.spec.ts

**Comment by watjurk** [comment]
I'm in progress of creating those test but when I try to run them I'm getting an error: ```Error: This environment's XHR implementation cannot support binary transfer```. This error repeats for every test case.

**Comment by johanbrandhorst** [comment]
Hm, I don't have much experience with the tests, but are you running it in node by any chance? Or, like, a really old browser?

**Comment by watjurk** [comment]
Good question, I assume they are running in node, but I expected this to work because I'm running `npm run start` in `integration_test` folder.

**Comment by watjurk** [comment]
I think I've resolved my issue. Test are running successfully now.

**Comment by johanbrandhorst** [comment]
Do you mean that the original issue is no longer present? If so could you please close the issue?

**Comment by watjurk** [comment]
No the original issue is still present, I mean that I know how to run your's tests successfully.

**Comment by watjurk** [comment]
Okay, I've just modified `cancellation.spec.ts` to `runWithSupportedTransports` and test is failing with websocket transport

**Comment by johanbrandhorst** [comment]
Thanks for the PR, but it looks like it's passing in CI?

**Comment by watjurk** [comment]
That's because tests never run in CI, see details in CI logs:
```
+ karma start ./karma.conf.ts --single-run
21.023633 testserver: 2021/03/10 21:56:10 Starting servers. http1.1 port: 9090, http1.1 empty port: 9095, http2 port: 9100, http2 empty port: 9105
23.214382 
23.214444 START:
23.238775 10 03 2021 21:56:12.631:INFO [karma-server]: Karma v6.0.3 server started at https://testhost:9876/
23.238950 10 03 2021 21:56:12.632:INFO [launcher]: Launching browsers Linux_chrome_41 with concurrency 4
23.249055 10 03 2021 21:56:12.641:INFO [launcher]: Starting browser Linux_chrome_41
27.561711 10 03 2021 21:56:16.955:INFO [Chrome 41.0.2272.76 (Linux i686)]: Connected on socket j5sxrhlc67zaPrVIAAAB with id 40222341
28.575147 Chrome 41.0.2272.76 (Linux i686) ERROR
28.575223   An error was thrown in afterAll
28.575240   Uncaught SyntaxError: Unexpected token )
28.575251   SyntaxError: Unexpected token )
28.578474 
28.578532 Finished in 0.107 secs / 0 secs @ 21:56:17 GMT+0000 (Coordinated Universal Time)
28.578542 
28.578618 SUMMARY:
28.578703 ✔ 0 tests completed
28.579476 
28.579533 ✖ Error while running the tests! Exit code: 1
28.579542 
28.579766 Chrome 41.0.2272.76 (Linux i686) ERROR
28.579776   Some of your tests did a full page reload!
28.580078 
28.580124 ✖ Error while running the tests! Exit code: 1
28.580134 
+ killGoTestServer
+ echo 'Killing testserver...'
+ killall testserver
31.919059 Killing testserver...
+ killTunnels
+ echo 'Killing Sauce Labs Tunnels...'
+ [[ -z '' ]]
+ kill 1204
+ [[ ! -z '' ]]
31.922051 Killing Sauce Labs Tunnels...
31.922156 10 Mar 21:56:21 - Got signal terminated
npm ERR! code 1
npm ERR! path /go/src/github.com/improbable-eng/grpc-web/integration_test
npm ERR! command failed
npm ERR! command sh -c ./run-with-tunnel.sh ./run-with-testserver.sh karma start ./karma.conf.ts --single-run

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/circleci/.npm/_logs/2021-03-10T21_56_21_322Z-debug.log
32.571899 10 Mar 21:56:21 - Cleaning up the tunnel connection to maki1639.miso.saucelabs.com
32.571969 10 Mar 21:56:21 - Cleaning up.
32.572073 10 Mar 21:56:21 - Removing tunnel 7f29d07de32649f7b4791d1b7c52af21.
32.572118 10 Mar 21:56:21 - Waiting for any active jobs using this tunnel to finish.
32.572167 10 Mar 21:56:21 - Press CTRL-C again to shut down immediately.
32.572206 10 Mar 21:56:21 - Note: if you do this, tests that are still running will fail.
36.607968 10 Mar 21:56:26 - All jobs using the tunnel have finished.
36.608046 10 Mar 21:56:26 - Waiting for the connection to terminate...
37.608364 10 Mar 21:56:27 - Connection closed (8).
37.608428 10 Mar 21:56:27 - Goodbye.
```
I think CI is broken and needs a fix. If you would checkout on my branch and run tests locally you would see them failing.

**Comment by MarcusLongmuir** [comment]
Fixed in #917
